rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -110,64 +138,175 @@ public void close(ISourceLocation loc) {
+         * meantime (in which case the current request is abandoned)
+         * @param calculation the actual summary calculation
+         */
+        private void debounce(int version, AtomicInteger latestVersion, Duration delay, Runnable calculation) {
+
+            latestVersion.set(version);
+            // Note: No additional logic (`compareAndSet` in a loop etc.) is
+            // needed to change `latestVersion`, because:
+            //   - LSP guarantees that the client sends change and save
+            //     notifications in-order, and that the server receives them
+            //     in-order. Thus, the version number of a file monotonically
+            //     increases with each notifications to be processed.
+            //   - To process notifications, calls of `didChange` and `didSave`
+            //     in `ParametricTextDocumentService` run sequentially and
+            //     in-order; these are the only methods that (indirectly) call
+            //     `calculate`. Thus, parameter `version` (obtained from the
+            //     notifications) monotonically increases with each `calculate`
+            //     call.
+
+            var delayed = CompletableFuture.delayedExecutor(delay.toMillis(), TimeUnit.MILLISECONDS, exec);
+            CompletableFuture.runAsync(() -> {
+                // If no new call to `calculate` has been made after `delay` has
+                // passed (i.e., `lastVersion` hasn't changed in the meantime),
+                // then run the calculation. Else, abandon this calculation.
+                if (latestVersion.get() == version) {
+                    calculation.run();
+                }
+            }, delayed);
         }
 
-        private void reportTypeCheckerMessages(List<Diagnostic> messages) {
-            typeCheckerMessages = messages;
-            sendDiagnostics();
+        public void calculateAnalyzer(CompletableFuture<Versioned<ITree>> tree, int version, Duration delay) {
+            debounce(version, latestVersionCalculateAnalyzer, delay, () ->
+                analyzer.calculateSummary(tree).thenAcceptIfUninterrupted(messages ->
+                    reportDiagnostics(analyzerDiagnostics, version, messages.get())));
         }
 
-        public void invalidate(boolean isClosing) {
-            summary.invalidate(isClosing);
+        /**
+         * The main complication when running the builder is that it might
+         * produce a subset of the same diagnostics as the analyzer. Thus, as
+         * the latest parser/analyzer/builder diagnostics are always reported
+         * together, a diff needs to be computed of the analyzer diagnostics and
+         * the builder diagnostics to avoid reporting duplicates produced by
+         * both analyzer and builder.
+         */
+        public void calculateBuilder(CompletableFuture<Versioned<ITree>> tree) {
+
+            // Schedule the analyzer. This is *always* needed, because the
```

Comment:
this code has gotten simpler, good. I would have expected the `builderBuild` to be a local variable.